Enable support DllImport a native assembly whose name contains '.'

luqunl opened the pull request:
Currently, In Windows, if the name your native dll specified in DllImport contains '.', such as Test.Foo, DllImport will only search for file Test.Foo and it won't search for file Test.Foo.dll

The fix is to append ".dll" if the native dll name specified in DllImport doesn't contain ".dll" .

Fix #17150

jkotas commented:
Test?

danmosemsft commented:
Ideally a test in Corefx so we can ensure it is fixed in Mono, CoreRT.

jkotas commented:
CoreCLR has better setup for interop tests. And Mono and CoreRT can run CoreCLR tests too.

mikedn commented:
> The fix is to append ".dll" if the native dll name specified in DllImport doesn't contain ".dll" .

So what happens if the specified name is "foo.exe"? Will it search for "foo.exe.dll"?

kasper3 commented:
I totally missed your PR. This approach of using `DetermineLibNameVariations` for both Windows and Unix is most probably neater than what I was experimenting with https://github.com/kasper3/coreclr/commit/edb45878f6915bf7443d74f7da2d17dd1dae5ac9, except for missing `.exe` check.
Also, does it account for casing `.dll`, `.DLL` etc.?

jkotas commented:
@luqunl Could you please rebase away the commits not related to your change?

luqunl commented:
Let me try to rebase my changes

luqunl commented:
> Could you please rebase away the commits not related to your change?
Fixed

> Test?
Fixed

> So what happens if the specified name is "foo.exe"? Will it search for "foo.exe.dll"?
Thanks for catching this. Fixed

> Also, does it account for casing .dll, .DLL etc.?
Thanks for catching this. Fixed

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6042,20 +6052,26 @@ HMODULE NDirect::LoadFromNativeDllSearchDirectories(AppDomain* pDomain, LPCWSTR
             containsSuffix = it == libName.End() || *it == (WCHAR)'.';
         }
+#else
+        if (libName.EndsWithCaseInsensitive(PLATFORM_SHARED_LIB_SUFFIX_W) || libName.EndsWithCaseInsensitive(WINDOWS_SHARED_EXE_SUFFIX_W))
```
I do not understand why we are special casing `.exe` suffix here. `.exe` suffix is non-standard suffix for shared libraries. If I want to, I can use `.dat` suffix for my .dlls and LoadLibrary will load them too.

luqunl replied:
The only reason is to avoid to load/search .exe.dll file. if it is possible that someone name his file as a.exe.dll, then we can remove this special case.

jkotas replied:
I am more worried about that cases that used to work fine before this change, but do not work anymore after this change. Consider:

```csharp
[DllImport("mydll.dat")]
extern static int sum(int a, int b);

static void Main() {
    Console.WriteLine(sum(1,2).ToString());
}
```

And `mydll.dat` that is renamed `mydll.dll` to `mydll.dat`. It works fine before this change (`mydll.dat` is passed to LoadLibrary), but it does not work after this change (`mydll.dat.dll` is passed to LoadLibrary).

luqunl replied:
Okay. 
`
            static const char* const SuffixFirst[] =
             {
                 "%.0s%s%s", // name+suffix
#ifdef FEATURE_PAL
                 "%s%s%s",   // prefix+name+suffix
#endif
                 "%.0s%s",   // name
#ifdef FEATURE_PAL
                 "%s%s%.0s"  // prefix+name
#endif
             };
'

luqunl replied:
I can check // name first, then //name + suffix to be more comp with current behavior.

jkotas commented on tests/src/Interop/DllImportAttribute/Simple/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+add_library (DllFileProbe SHARED ${SOURCES})
+target_link_libraries(DllFileProbe ${LINK_LIBRARIES_ADDITIONAL}) 
+
+#get_cmake_property(_variableNames VARIABLES)
```
Do we need these commented out blocks?

luqunl replied:
I will remove these commented lines..just copy & paste..

luqunl replied:
Removed these comments

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6078,9 +6096,13 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
             static const char* const SuffixFirst[] =
             {
                 "%.0s%s%s", // name+suffix
```
For name without any dots, this is going to probe twice now:

- First time, we append the `.dll`
- Second time, we pass in name without `.dll` suffix and LoadLibrary appends it internally.

It would be nice to avoid this duplication (e.g. just let LoadLibrary take care of appending .dll to the simple names).

luqunl replied:
Fixed

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6078,9 +6096,13 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
             static const char* const SuffixFirst[] =
             {
                 "%.0s%s%s", // name+suffix
+#ifdef FEATURE_PAL
```
Similar for names that end with a `.`. We should not do any probing for them on Windows - to maintain the behavior documented for them on MSDN.

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -5890,6 +5890,16 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
+#else // !FEATURE_PAL
+#define PLATFORM_SHARED_LIB_SUFFIX_A ".dll"
+#define PLATFORM_SHARED_LIB_PREFIX_A ""
+#define PLATFORM_SHARED_LIB_SUFFIX_W W(".dll")
```
Nit: I do not think we need to have a defines for these since they are used just once.

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6118,6 +6104,36 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
         *numberOfVariations = COUNTOF(NameOnly);
     }
 }
+#else
```
Nit: It would be nice to add `// FEATURE_PAL` comment here.

luqunl replied:
Updated

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6118,6 +6104,36 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
+        *numberOfVariations = COUNTOF(SuffixLast);
+    }
+}
+#endif
```
Same here.

luqunl replied:
Updated

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6118,6 +6104,36 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
+    bool containsSuffix = libName.EndsWithCaseInsensitive(W(".dll")) || libName.EndsWithCaseInsensitive(W(".exe"));
+    if (endWithDot || notContainDot || containsSuffix || !libNameIsRelativePath)
+    {
+        // file name end with '.'
```
This comment is not very useful. It is a literal rewrite of the above condition in plain English.

Instead, it would be useful to have comment that describes the reasons for this logic.

luqunl replied:
Added more comments

jkotas commented:
Build errors:

```
src\vm\dllimport.cpp(6110): error C2220: warning treated as error - no 'object' file generated 
src\vm\dllimport.cpp(6110): warning C4800: 'BOOL': forcing value to bool 'true' or 'false' (performance warning)
```

luqunl commented:
Fixed build break for BOOL to bool conversation.

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6096,7 +6104,42 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
+    bool containsSuffix = libName.EndsWithCaseInsensitive(W(".dll")) || libName.EndsWithCaseInsensitive(W(".exe"));
+    
+    // The purpose of following code is to workaround LoadLibrary limitation: 
+    // LoadLibray won't append extension if filename itself contains '.'. Thus It will break the following scenario:
```
Typo:  It -> it

luqunl replied:
Fixed

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6096,7 +6104,42 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
+#else // FEATURE_PAL
+static void DetermineLibNameVariations(const char* const** libNameVariations, int* numberOfVariations, const SString& libName, bool libNameIsRelativePath)
+{
+    bool endWithDot = libName.EndsWith(W(".")) == TRUE;
```
It is a bad practice to compare `BOOL` to `TRUE`. `BOOL` is actually integer and any non-zero value can mean true. `TRUE` is defined as `1`. So comparing `BOOL` to `TRUE` assumes that the non-zero true value is `1` that is not always the case.

Use either `!!libName.EndsWith(W("."))`  or `libName.EndsWith(W(".")) != FALSE`.

luqunl replied:
Thanks for this. Previous I had searched the implementation for SString.EndWith and it always return TRUE or FALSE.  I will use !! instead.

luqunl commented:
> tests/src/Interop/DllImportAttribute/FileExtensionProbe.cs is encoded as UTF16LE (file --mime-encoding FileExtensionProbe.cs), and github is showing it as a binary file in diff view. You can fix it by changing it to ASCII using notepad file>saveAs or bash: iconv -f utf-16le -t US_ASCII FileExtensionProbe.cs > FileExtensionProbe.cs_new then delete the old and rename _new one.

Thanks a lot for this info. I saw this issue but didn't realize that this is file encoding issue.

luqunl commented:
> Typo in loadlibrary
Fixed

kasper3 commented:
@jkotas, is there a chance to include this in 2.1 release?

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6096,7 +6104,42 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
+#else // FEATURE_PAL
+static void DetermineLibNameVariations(const char* const** libNameVariations, int* numberOfVariations, const SString& libName, bool libNameIsRelativePath)
+{
+    bool endWithDot = !!libName.EndsWith(W("."));
```
Now that I am looking at this again ... caching these in bool variables makes this less efficient than it needs to be. If you put these directly into the if statement without caching them in locals like this:

```
if (!libName.Find(libName.Begin(), W('.')) ||
    libName.EndsWith(W(".")) ||
    libName.EndsWithCaseInsensitive(W(".dll")) || libName.EndsWithCaseInsensitive(W(".exe"))
{
```

It will be more efficient because of it will stop at `!libName.Find(libName.Begin(), W('.'))` for names without dot; and the remaining two statements won't be executed at all.

It also avoids the `!!` wart.

jkotas commented:
> @jkotas, is there a chance to include this in 2.1 release?

@kasper3 Why do you think it is important to include this in 2.1 release?

kasper3 commented:
> There would need to be a good justification why this change is important to include in 2.1

In 2.0 and older versions, we are using build time conditions:

```c#
#if WINNT
[DllImport("shared.library.dll")]
#else
[DllImport("shared.library")]
#endif
```

With this change, it will be reduced to `[DllImport("shared.library")]`. The sooner this inconsistency goes away, the better.

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6096,7 +6104,42 @@ static void DetermineLibNameVariations(const char* const** libNameVariations, in
+        libName.EndsWith(W(".")) || 
+        libName.EndsWithCaseInsensitive(W(".dll")) || 
+        libName.EndsWithCaseInsensitive(W(".exe")) || 
+        !libNameIsRelativePath)
```
`libNameIsRelativePath` should be first because of it is the cheapest condition to evaluate.

jkotas replied:
Also, is the relative path condition really correct here? Relative path is going to be for a path like:

`XXX\YYY\library`

or

`\XXX\YYY\library`.

On Unix, we will end up trying paths like:

`libXXX\YYY\library.so`

or 

`lib\XXX\YYY\library`.

luqunl replied:
Move libNameIsRelativePath  to the first one. 

 'is the relative path condition really correct here?

This is incorrect. since this is an old issue and I will create another issue to track and solve this problem

jkotas commented on src/vm/dllimport.cpp:
```diff
@@ -6175,11 +6217,7 @@ HINSTANCE NDirect::LoadLibraryModule(NDirectMethodDesc * pMD, LoadLibErrorTracke
-    {
-        LPCWSTR currLibNameVariation = wszLibName;
-#endif
+        currLibNameVariation.Printf(prefixSuffixCombinations[i], PLATFORM_SHARED_LIB_PREFIX_A, name, PLATFORM_SHARED_LIB_SUFFIX_A);
```
We have converted the name from UTF8 to Unicode above once already.

The PrintF is going to convert it again; but I do not think it is going to use UTF8 on Windows. I think this needs to use the converted name to Unicode (wszLibName); not the UTF8 name.

luqunl replied:
Update all of these argument to WCHAR* instead of char*

jkotas commented:
> include this in 2.1 release

There seems to be way too many subtle issues in this change. I do not think we should be considering this for 2.1 release.

jkotas commented:
LGTM

jkotas commented:
master is open for 2.2 now. Merge means that we believe it is a good change for 2.2.

It would be up to @luqunl to present it to the shiproom to consider it for 2.1. My guess is that it would not meet the bar.

sdmaclea commented:
This file introduced corrupt line endings https://github.com/dotnet/coreclr/pull/17505/files#diff-23d6cee28d9638de3cc99b9da9fbb3b9
